app/domain/authentication/authn_k8s/application_identity.rb

```diff
@@ -0,0 +1,181 @@
+module Authentication
+  module AuthnK8s
+
+    Err = Errors::Authentication::AuthnK8s
+    # Possible Errors Raised: MissingNamespaceConstraint, IllegalConstraintCombinations
+    # ScopeNotSupported, ArgumentError
+
+    # This class defines a application identity of a given conjur host.
+    # The constructor initializes an ApplicationIdentity object and validates that
+    # it is configured correctly.
+    # The difference between the validation in this constructor and
+    # the validation in ValidateApplicationIdentity is that here we validate that
+    # the application identity is configured correctly, and thus is a valid application
+    # identity. In ValidateApplicationIdentity we validate that the defined application
+    # identity is actually the correct one in kubernetes.
+    # For example, a application identity `some-namepsace/blah/some-value` is not
+    # valid and will fail the validation here. However, a application identity
+    # `some-namespace/service-account/some-value` is a valid application identity
+    # and will pass the validation here. If the host is actually running from
+    # a pod with service account `some-other-value` then it will fail the
+    # validation of ValidateApplicationIdentity
+    class ApplicationIdentity
+
+      def initialize(host_id:, host_annotations:, service_id:)
+        @host_id          = host_id
+        @host_annotations = host_annotations
+        @service_id       = service_id
+
+        validate
+      end
+
+      def namespace
+        @namespace ||= application_identity_in_annotations? ? constraint_from_annotation("namespace") : host_id_suffix[-3]
+      end
+
+      def constraints
+        @constraints ||= {
+          pod:               constraint_value("pod"),
+          service_account:   constraint_value("service_account"),
+          deployment:        constraint_value("deployment"),
+          deployment_config: constraint_value("deployment_config"),
+          stateful_set:      constraint_value("stateful_set")
+        }
+      end
+
+      def container_name
+        annotation_name = "authentication-container-name"
+        annotation_value("authn-k8s/#{@service_id}/#{annotation_name}") ||
+          annotation_value("authn-k8s/#{annotation_name}") ||
+          annotation_value("kubernetes/#{annotation_name}") ||
+          "authenticator"
+      end
+
+      def namespace_scoped?
```

Comment:
Is this method used to determine if we use the old style vs new style for lookup?

Reply:
no. It is used to determine if the application identity has a namespace granularity. For example:
```
    - !host
      id: test-app-namespace
      annotations:
        authn-k8s/namespace: {{CONJUR_AUTHN_K8S_TEST_NAMESPACE}}
```
is namespace scoped while
```
    - !host
      id: test-app-service-account
      annotations:
        authn-k8s/namespace: {{CONJUR_AUTHN_K8S_TEST_NAMESPACE}}
        authn-k8s/service_account: inventory-pod-only
```
is not.

Same goes for the old style - the host `.../test-namespace/*/*` is namespace scoped while `.../test-namespace/service_account/test-sa` is not.

Reply:
I added a comment in the code:
```
      # returns true if the only constraint is on the namespace, false otherwise
      def namespace_scoped?
        constraints.compact.empty?
      end
```